docs/proposals/addon-auth-pattern.md

```diff
@@ -0,0 +1,111 @@
+
+This document proposes a reference pattern for cluster addons to authenticate users, and suggests two new 
+features that enable this pattern.
+
+### Goals
+
+- Programs which run on the kubernetes cluster, and which provide services to users of the cluster,
+  can authenticate those users.
+  - Can implement this without the need to write much authentication code, allowing ease of development/extending kubernetes.
+- Programs that do not need to handle authentication credentials do not have to, thus allowing a "service admin"
+  to maintain a service, without that service admin ever getting access to other users authentication creds,
+  which could then be replayed against other services.
+- To keep authentication and authorization separate, since this improves composability.
+- Allow users to reuse the cluster credentials in their `~/.kube/config`, so they do not need
+  creds per cluster
+- Do not use basic-auth creds in browsers, which are subject to CSRF attacks.
+- Do not use client certs in browsers, which are hard to install.
+
+The example used in this pattern is the kubernetes Dashboard.
+
+### Current pattern which has problems
+
+The following pattern is in use today.
+
+1. Dashboard is running with service `dashboard` in namespace `kube-system`.
+1. User runs `kubectl proxy -p 8001 &`.
+  -  This causes kubectl to serve HTTP on `localhost:8001`, and to reverse proxy that to the users current cluster, using TLS and the users
+     currently configured authentication method.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/#/workload`
+  - This causes the API server to proxy a single HTTP request.  The effect of the two proxies is that an HTTP request on localhost:8001 becomes an HTTP request for `/workloads` on port 80 on the pod IP of one of the Pods that backs services `dashboard` in namespace `kube-system`.  So far so good.
+1. The response is sent back from the Pod to the Apiserver.
+1. The response is sent back to the users.
+1. The user sees the Kubernetes Dashboard.
+
+The problem with this is that the Dashboard does not know which user is accessing it, so it cannot support multiple users with different capabilities.
+
+### Proposed Pattern
+
+The proposed pattern has the following key improvements over the current pattern:
+- a header is injected which tells the Dashboard pod the user who is reaching it.
+- the header can be validated easily as being signed by the apiserver, which is trusted to authenticate users
+
+1. Dashboard is running with service `dashboard` in namespace `kube-system`.
+1. User runs `kubectl proxy -p 8001 &`.
+  -  Same as before.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/?add-auth=true#/workload`
+  - Difference is the addition of the `add-auth=true` query parameter, which is handled by the APIserver.
```

Comment:
Good point, what I said won't work.  I've changed this to not use a query parameter, but instead use an annotation on the service.  This makes more sense than requiring the user to specify it, since the need for authentication headers is a property of the service.